File: app/src/main/java/com/theandroiddev/mywins/SuccessImageAdapter.java
package com.theandroiddev.mywins;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.constraint.ConstraintLayout;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.util.List;

import static android.content.ContentValues.TAG;

/**
 * Created by grazyna on 2017-08-23.
 */

class SuccessImageAdapter extends RecyclerView.Adapter<SuccessImageAdapter.ViewHolder> {

    private List<SuccessImage> successImages;
    private OnSuccessImageClickListener listener;
    private int successImageLayout;


    SuccessImageAdapter(List<SuccessImage> successImages, OnSuccessImageClickListener listener, int successImageLayout) {
        this.successImages = successImages;
        this.listener = listener;
        this.successImageLayout = successImageLayout;
    }

    @Override
    public SuccessImageAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(successImageLayout, parent, false);
        return new SuccessImageAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(SuccessImageAdapter.ViewHolder holder, int position) {

        if (successImages.get(position).getImagePath() == null) {
            if (position == 0) {
                holder.successImageIv.setImageResource(R.drawable.ic_action_add);
                holder.bind(successImages.get(position), position, listener);

            }
        } else {

            Bitmap thumbnail = getbitpam(successImages.get(position).getImagePath());
            Bitmap orientedBitmap = ExifUtil.rotateBitmap(successImages.get(position).getImagePath(), thumbnail);
            holder.successImageIv.setImageBitmap(orientedBitmap);
            holder.bind(successImages.get(position), position, listener);

        }

    }

    private Bitmap getbitpam(String path) {
        Bitmap imgthumBitmap = null;
        try {

            final int THUMBNAIL_SIZE = 256;

            FileInputStream fis = new FileInputStream(path);
            imgthumBitmap = BitmapFactory.decodeStream(fis);

            imgthumBitmap = Bitmap.createScaledBitmap(imgthumBitmap,
                    THUMBNAIL_SIZE, THUMBNAIL_SIZE, false);

            ByteArrayOutputStream bytearroutstream = new ByteArrayOutputStream();
            imgthumBitmap.compress(Bitmap.CompressFormat.JPEG, 100, bytearroutstream);


        } catch (Exception ex) {
            Log.d(TAG, Log.getStackTraceString(ex));

        }
        return imgthumBitmap;
    }


    @Override
    public int getItemCount() {
        return successImages.size();
    }

    interface OnSuccessImageClickListener {
        void onSuccessImageClick(SuccessImage successImage, ImageView successImageIv, int position, ConstraintLayout constraintLayout,
                                 CardView cardView);

        void onSuccessImageLongClick(SuccessImage successImage, ImageView successImageIv, int position, ConstraintLayout constraintLayout,
                                     CardView cardView);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        ImageView successImageIv;
        ConstraintLayout constraintLayout;
        CardView cardView;


        ViewHolder(View itemView) {
            super(itemView);

            successImageIv = itemView.findViewById(R.id.success_image_iv);
            constraintLayout = itemView.findViewById(R.id.success_image_constraint_layout);
            cardView = itemView.findViewById(R.id.success_image_card_view);

        }

        private void bind(final SuccessImage successImage, final int position, final OnSuccessImageClickListener onClickListener) {

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onSuccessImageClick(successImage, successImageIv, position, constraintLayout, cardView);
                }
            });
            itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    listener.onSuccessImageLongClick(successImage, successImageIv, position, constraintLayout, cardView);
                    return true;
                }
            });
        }
    }
}

File: app/src/main/java/com/theandroiddev/mywins/presentation/image/SuccessImageAdapter.kt
package com.theandroiddev.mywins.presentation.image

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.cardview.widget.CardView
import androidx.constraintlayout.widget.ConstraintLayout
import com.squareup.picasso.Picasso
import com.theandroiddev.mywins.R
import com.theandroiddev.mywins.presentation.image.SuccessImageAdapter.ViewHolder
import com.theandroiddev.mywins.presentation.successes.SuccessImageModel
import java.io.File

class SuccessImageAdapter(
    private val listener: OnSuccessImageClickListener,
    private val successImageLayout: Int
) : androidx.recyclerview.widget.RecyclerView.Adapter<ViewHolder>() {

    var successImages: MutableList<SuccessImageModel> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(successImageLayout, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        if (successImages[position].imagePath.isEmpty()) {
            if (position == 0) {
                holder.successImageIv.setImageResource(R.drawable.ic_action_add)
                //holder.cardView.setCardBackgroundColor(context.getResources().getColor(R.color.primary));
                holder.bind(successImages[position], position)

            }
        } else {

            Picasso.get().load(File(successImages[position].imagePath)).resize(256, 256).centerCrop().into(holder.successImageIv)
            holder.bind(successImages[position], position)

        }

    }

    override fun getItemCount(): Int {
        return successImages.size
    }

    interface OnSuccessImageClickListener {
        fun onSuccessImageClick(successImage: SuccessImageModel, successImageIv: ImageView, position: Int, constraintLayout: ConstraintLayout,
                                cardView: CardView
        )

        fun onSuccessImageLongClick(successImage: SuccessImageModel, successImageIv: ImageView, position: Int, constraintLayout: ConstraintLayout,
                                    cardView: CardView)
    }

    inner class ViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {

        var successImageIv: ImageView
        var constraintLayout: ConstraintLayout
        var cardView: CardView


        init {

            successImageIv = itemView.findViewById(R.id.success_image_iv)
            constraintLayout = itemView.findViewById(R.id.success_image_constraint_layout)
            cardView = itemView.findViewById(R.id.success_image_card_view)

        }

        fun bind(successImage: SuccessImageModel, position: Int) {

            itemView.setOnClickListener { listener.onSuccessImageClick(successImage, successImageIv, position, constraintLayout, cardView) }
            itemView.setOnLongClickListener {
                listener.onSuccessImageLongClick(successImage, successImageIv, position, constraintLayout, cardView)
                true
            }
        }
    }

    fun addSuccessImage(position: Int, successImage: SuccessImageModel) {
        this.successImages.add(position, successImage)
    }

    fun updateSuccessImage(position: Int, successImage: SuccessImageModel) {
        this.successImages[position] = successImage
    }

    fun addSuccessImage(successImage: SuccessImageModel) {
        this.successImages.add(successImage)
    }
}
